### ISSUE TITLE
missing renavigate method at new angular-router

### ISSUE THREAD

**Issue by gustavolira** [angular/angular#9669]
Hi,

In angular/router-deprecated we have Router.renavigate(), but at new component of router, I mean angular/router does not have something like Router.renavivate().
Would be nice if we have the same behavior a angular/router.

Thanks

**Comment by zoechi** [comment]
Do you have a concrete use case?

**Comment by gustavolira** [comment]
@zoechi It's simple, just look for Router at angular/router-deprecated package and you will see renavigate method

**Comment by zoechi** [comment]
@gustavolira I know it was there but some things are working differently in the new router and perhaps your concrete use case could be accomplished with existing features, and if not it would still be interesting to know your use case to ensure renavigate does what you expect, otherwise it might be sufficient to add a `renavigate()` method with an empty body ;-)

**Comment by PEsteves8** [comment]
Personally I wanted this, because when I use canActivate with a guard, sometimes instead of redirecting to somewhere specific I would prefer to keep the exact same url. If I just return false, the user stays in the same place as expected, but the url changes stays as whatever was inserted (the unaccessible url)

**Comment by awerlang** [comment]
@PEsteves8 isn't this a bug then? the route url should reflect the current state in the outlet's.

**Comment by PEsteves8** [comment]
Well, if the point is to have the url stay what it was before the blocked route access attempt, then I suppose it it.

**Comment by janerist** [comment]
My use case for this is an edit form where you want to reset the form state after saving, and also make sure that your form is up to date with data from the server. Sure, I can use router.navigate(...) and set up an observable and react to changed data and update my form that way, but that introduces a lot of unnecessary complexity when just reloading the current route would achive the same thing a lot simpler.

**Comment by damiandennis** [comment]
same problem here, why cant we renavigate anymore? it was a quick solution to reloading all state of a path. Yes I can use observables but then everything has to be hooked up to observables and with a lot of interactions on a page this can get difficult to manage. A simple renavigate made this simple. I think the method reload makes more sense as well.

**Comment by Krisa** [comment]
Similarly, the point of renavigate is to avoid all kind of boilerplate. I have an edit form, which can be called from multiple places in the application. Using renavigate when the user clicks save (or cancel!) redirects exactly to the place where the user was before. It's possible now to do the same, somehow, but it's (much) more complicated, etc.

**Comment by zoechi** [comment]
@Krisa the new forms got a `reset()` method recently AFAIK

**Comment by Krisa** [comment]
on the FormGroup or FormControl, something like this https://github.com/angular/angular/blob/da8eb9f8b87a446a57a5ca006cb15c7b70828918/modules/%40angular/forms/test/model_spec.ts#L773 ? 

This is not yet in the current RC4, but more particularly, my use case is actually not really about that. I'm looking for a way to navigate back to the last visited url within the application. Kind of history.back() but within Angular2 mechanism (and additionally I don't want to navigate back in the history, but to move forward even so the url will be the same).

**Comment by roben** [comment]
Same for me. I have a rather complex application which components react on series of global events. If the event history is changed (i.e. undo), it is waay more complicated to do atomic undos than just reload everything with the new event history.

Currently i am working around this by simply calling `location.reload();`

**Comment by QuentinFchx** [comment]
I also have the following use-case:
I have an app where you can switch between several "accounts".
Each accounts has different permission, and some routes should be inaccessible to some accounts.

Let's say userA can access routeA, but not userB.
The routeA is protected by a guard redirecting to routeX any unauthorized user.

If I'm on routeA as userA, and I switch to userB, I would like to re-run the guards stack and redirect userB to routeX.

**Comment by kblestarge** [comment]
What is the status of this feature request?

**Comment by msieurtoph** [comment]
Same usecase as @QuentinFchx ...
I do not understand why `renavigate()` has been removed.
`location.reload()` seems to me to be a (way too much) radical solution.

**Comment by flensrocker** [comment]
My usecase is the same as @janerist (reload an edit-form after save to reflect the new data).

The form creation is hooked up to the `route.params` observable, but now I do a `combineLatest` with some `BehaviorSubject<boolean>(true)` to trigger an refresh with a call to its `next` method. That makes the parameters of the `map` of the `Params` a lot uglier.

If someone has a nicer way of reloading, please let me know...

```javascript
export class MyComponent {
  public form: Observable<FormGroup>;
  private _reload: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(true);

  constructor(route: ActivatedRoute) {
    this.form = route.Params
      .combineLatest(this._reload)
      .map((data: [Params, boolean]) => {
        let params: Params = data[0];
        ...
      });
  }

  renavigate(): void {
    this._reload.next(true);
  }
}
```

**Comment by intellix** [comment]
My use-case: Route has an AuthGuard and when you click Logout in the header, you'd like to re-run guards to know if the user should be on his current page. You need to re-run the guard and currently it requires repeating the code here to find guards, inject and run them: https://github.com/angular/angular/blob/8f7915022c61b86194bebce9ce6591c7ced40551/packages/router/src/pre_activation.ts#L291

Not saying we need a renavigate API exposed, but perhaps the Guard running API should be exposed so we can just tell them to re-run

**Comment by cwalcott** [comment]
The use case I have are components with resolved data. Sometimes I want the resolved data to be reloaded in response to a user action.

For example, a component that lists users and includes the ability to delete a user. After a user is deleted, I want the component to receive an updated set of users without the deleted one.

For now I've been using something similar to "Workaround 1" in [intellix's comment](https://github.com/angular/angular/issues/9669#issuecomment-330163462) though it requires incrementing the "reload" parameter if it exists (to handle multiple refreshes).

**Comment by awerlang** [comment]
@cwalcott This is already supported by resolvers. Resolver data is available in the `.data` field as an observable as you should be already aware. Don't take only the first item, keep a subscription to it. When it's time to reload, make the resolver `.next()` another item. Of cource it takes some more work, but it doesn't need any workaround to do so, it's how it is designed.

**Comment by cwalcott** [comment]
@awerlang Don't observables from resolvers have to complete before the component gets created? I tried making a resolver that reloads items based on mapping over a subject, but the component never loads, I think because the resolver's observable never completes:

```typescript
@Injectable()
export class UsersResolver implements Resolve<User[]> {
  private reloadSubject = new BehaviorSubject(true);

  constructor(private backendService: BackendService) {}

  reload(): void {
    this.reloadSubject.next(true);
  }

  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<User[]> {
    return this.reloadSubject.switchMap(() => this.backendService.getUsers())
  }
}
```

**Comment by awerlang** [comment]
@cwalcott you're right, sorry for misleading you.
For CanActivate it was the same thing, until I sent a PR to fix #9613. 

So what I described earlier is how I pictured it in my mind. I think it needs to work that way. I fail to understand why consume data as an observable in the current approach, it could be the value itself. Probably I'm missing something. There weren't much prior art for observable-based routers, so this must have been a slip in the design.

I found this PR #19101, although I don't think it's a good solution for resolve (i.e. take the first value).

**Comment by blahblehblah** [comment]
Any update on this?

**Comment by flensrocker** [comment]
Another workaround we're using, is
```typescript
import { Component } from '@angular/core';
import { Router, ActivatedRoute } from '@angular/router';

@Component({
    template: '<p>redirecting...</p>'
})
export class RedirectComponent {
    private returnUrl: string;

    constructor(
        private _router: Router,
        private _route: ActivatedRoute,
    ) {
        if (this._route.snapshot.paramMap.has("returnUrl")) {
            this.returnUrl = decodeURIComponent(this._route.snapshot.paramMap.get("returnUrl"));
        }
    }

    ngOnInit() {
        if (this.returnUrl) {
            this._router.navigateByUrl(this.returnUrl, { skipLocationChange: true });
        }
    }
}

...

const appRoutes: Routes = [
    { path: "redirect", component: RedirectComponent },
];

...

this._router.navigate(["/redirect", { returnUrl: this._router.url }], { skipLocationChange: true });
```
It's usally so fast, that you don't see the content of the template.

**Comment by mlc-mlapis** [comment]
... there is `onSameUrlNavigation` for that ... you can read https://github.com/angular/angular/issues/21115

**Comment by alex7egli** [comment]
Any updates on this? I have a global state that affects all the data on my current route, and if the state changes then all components need to refresh all their data. It would be a huge burden to add code to every single component to listen for this state change and reload everything, as opposed to how it worked in angularJs where I could just do $route.reload().

**Comment by mlc-mlapis** [comment]
@alex7egli ... do you expect creating new components instances?

**Comment by mlc-mlapis** [comment]
@alex7egli ... can you describe more closely how do you pass that global state to those components ... because it's unclear for me ... related to your point of view for the next re-initialization.

**Comment by alex7egli** [comment]
@mlc-mlapis I have a local data cache and the various components query the cache for data as needed, based on the component. On initialization most components check a global service for the current state. I don't want to go into too much detail because it isn't relevant to this issue, which is that Angular Router should have a reload type method that reloads the current route.

**Comment by alex7egli** [comment]
@mlc-mlapis So if you don't want to implement this feature because you can't understand the need for it, why is this ticket in the backlog and not closed? It's not like devs stuck using Angular aren't used to the response "Angular just doesn't work that way, switch to React".

**Comment by mlc-mlapis** [comment]
@alex7egli ... I was talking about the current state. As I understand ... there are some priorities (as Ivy renderer #1 actually) and the final decision has not yet happened ... so this feature issue is waiting for it ...
`.renavigate()` method and also `.reload()` (simplify programmatic reloading of the same route ... which keeps the component instance but could re-run its guards and resolvers).

**Comment by MatthiasKunnen** [comment]
I've updated @intellix's workaround for Angular 7
```TypeScript
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { ActivatedRoute, CanActivate, Router } from '@angular/router';

import { from, Observable, of } from 'rxjs';
import { mergeMap, takeWhile } from 'rxjs/operators';

// Todo Update when https://github.com/angular/angular/issues/9669 is resolved.
@Injectable()
export class RouteService {
    constructor(
        private injector: Injector,
        private route: ActivatedRoute,
        private router: Router,
    ) {
    }

    recursivelyGetGuardedRoutes(route: ActivatedRoute): ActivatedRoute[] {
        const config = route.routeConfig;

        // Check if current route has a canActivate guard
        const result = config && config.canActivate && config.canActivate.length > 0
            ? [route]
            : [];

        if (!route.children) {
            return result;
        }

        return result.concat(route.children
            .reduce((acc: Array<ActivatedRoute>, r: ActivatedRoute) => {
                return acc.concat(this.recursivelyGetGuardedRoutes(r));
            }, []));
    }

    /**
     * This returns an observable which will run all the canActivate guards of
     * the activated route and its children. The observable emits the first
     * false value or true.
     * @returns {Observable<boolean>}
     */
    runGuardChain(): Observable<boolean> {
        const routes = this.recursivelyGetGuardedRoutes(this.route);
        const obs = routes.reduce((acc: (Observable<boolean> | Promise<boolean>)[],
                                   route: ActivatedRoute) => {
            if (!route.routeConfig || !route.routeConfig.canActivate) {
                return acc;
            }

            return acc.concat(route.routeConfig.canActivate
                .map((token: InjectionToken<CanActivate>) => {
                    const guard = this.injector.get(token);
                    const value = guard.canActivate(
                        route.snapshot,
                        this.router.routerState.snapshot,
                    );

                    return typeof value === 'boolean' ? of(value) : value;
                }));
        }, []);

        // Loop through each guard whilst they return thruthy
        return from(obs).pipe(
            mergeMap(o => o),
            takeWhile(v => !!v),
        );
    }
}
```

**Comment by gund** [comment]
This feature is really required for the next use case:

Suppose we have a subsection (a feature) in app that describes RealEstate and everything related to it.

So the url will be: `/realestate/:id`.

We will have a `RealEstateResolver` that will sit on path `/realestate/:id` and fetch data from network.

Then we have multiple subsections where we:
1. `/realestate/:id/data` - Display and edit RE data
2. `/realestate/:id/other` - Display something relevant to RE (ads, brokers, etc.)

But when we edit our RE the new data will not be reflected across the sections because the resolver is stuck and there is no way to refresh it (apart from going briefly outside of `/realestate/:id` path and then back).

Also note that none of `runGuardsAndResolvers` options can help here - since we do not want to refetch RE between every sections navigation, but only when we do specific action (editing of RE).

This is the perfect use-case of `Router#renavigate()` method.

**Comment by pauldraper** [comment]
There are several ways to workaround this.

One possible way:

```ts
export class Reloadable<T> {
  private readonly reload$ = new BehaviorSubject(undefined);

  readonly value$ = this.reload$.pipe(switchMap(this.f));

  constructor(private readonly f: () => Observable<T>)  {}

  reload() {
    this.reload$.next(undefined);
  }
}

@Injectable()
export class MyResolver implements Resolve<Reloadable<number>> {
  resolve(route, state) {
    const reloadable = new Reloadable(() => of(Date.now()).pipe(delay(1000)));
    return reloadable;
    // or, if you need the value to exist immediately the first time
    // return reloadable.value.pipe(mapTo(reloadable));
  }
}

@Injectable()
export class MyConsumer implements OnDestroy {
  private readonly destroy = new Subject();

  readonly value = this.route.data.pipe(switchMap(data => data.get('key').value$));
  
  constructor(private readonly route: ActivatedRoute) {
    someChange$.pipe(withLatestFrom(this.route.data), takeUntil(this.destroy)).subscribe(([_, data]) => data.get('key').reload());
  }

  ngOnDestroy() {
    this.destroy.next();
  }
}
```

**Comment by theLeftTenant** [comment]
I like the resolver/guard concept, but without being able to reevaluate after conditions change, you're at a loss in common scenarios such as details-edit.

What if we could control this declaratively in the route config, which seems most conventional, instead of with the imperative AngularJS-style `router.reload()`?

```typescript
{
    path: 'widgets/:id',
    resolve: { widget: WidgetResolver },
    children: [
        {
            path: 'details',
            component: WidgetDetailsComponent,
            children: [
                {
                     path: 'gizmos',
                     component: GizmosPickerComponent,
                     rerunResolver: {
                         atPath: '../../', // relative or absolute
                         upon: 'leave'     // 'leave' || 'enter' etc.
                     }
                }
            ]
        },
        {
            path: 'edit',
            component: WidgetFormComponent,
            rerunResolver: {
                atPath: '../',  // relative or absolute
                upon: 'leave'   // 'leave' || 'enter' etc.
            }
        }
    ]
}
```

**Comment by atscott** [comment]
I believe this could be resolved if #23152 were merged, right? You would also need to provide a custom [RouteReuseStrategy](https://angular.io/api/router/RouteReuseStrategy) if you wanted components to reload (i.e.  `shouldReuseRoute` returning `false`).

**Comment by atscott** [comment]
Closing since it's unclear to me exactly what this feature request is for since the available configurations and features in the Router have changed significantly since 2016. The relevant items in the router could be [onSameUrlNavigation](https://angular.io/api/router/ExtraOptions#onSameUrlNavigation), [RouteReuseStrategy](https://angular.io/api/router/RouteReuseStrategy), [RunGuardsAndResolvers](https://angular.io/api/router/Route#runGuardsAndResolvers), returning navigation extras when returning a `UrlTree` from a guard (#27148), or just simply triggering a new navigation from within a guard/resolver. 

If there is no existing feature or feature request that covers some needed use-case here (potentially reloading on an individual navigation call as was in #23152), please file a new issue so it can be re-prioritized.
